dodiscover/toporder/_base.py

```diff
@@ -0,0 +1,631 @@
+    pns_threshold: float, optional
+        Threshold to use for PNS, default is 1.
+
+    Attributes
+    ----------
+    graph_ : nx.DiGraph
+        Adjacency matrix representation of the inferred causal graph.
+    order_ : List[int]
+        Topological order of the nodes from source to leaf.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(
+        self,
+        cam_cutoff: float = 0.001,
+        n_splines: int = 10,
+        splines_degree: int = 3,
+        pns: bool = False,
+        pns_num_neighbors: Optional[int] = None,
+        pns_threshold: float = 1,
+    ):
+        self.cam_cutoff = cam_cutoff
+        self.n_splines = n_splines
+        self.degree = splines_degree
+        self.do_pns = pns
+        self.pns_num_neighbors = pns_num_neighbors
+        self.pns_threshold = pns_threshold
+
+        self.order_: nx.DiGraph = nx.empty_graph()
+        self.graph_: List[int] = list()
+
+    def get_leaf(self, leaf: int, remaining_nodes: List[int], current_order: List[int]) -> int:
+        """Get leaf node from the list of `remaining_nodes` without an assigned order.
+
+        Parameters
+        ----------
+        leaf : int
+            Leaf position in the list of `remaining_nodes`.
+        remaining_nodes : List[int]
+            List of nodes without a position in the order.
+        current_order : List[int]
+            Partial topological order.
+
+        Returns
+        -------
+        leaf : int
+            Leaf index in the list of graph nodes.
+        """
+        # descendants enforced by edges in self.context and not in the order are used as leaf
+        leaf_descendants = self.order_constraints[remaining_nodes[leaf]]
+        if not set(leaf_descendants).issubset(set(current_order)):
+            k = 0
+            while True:
+                if leaf_descendants[k] not in current_order:
+                    leaf = remaining_nodes.index(leaf_descendants[k])
+                    break  # exit when leaf is found
+                k += 1
+        return leaf
+
+    def fit(self, data_df: pd.DataFrame, context: Context) -> None:
+        """
+        Fit topological order based causal discovery algorithm on input data.
+
+        Parameters
+        ----------
+        data_df : pd.DataFrame
+            Datafame of the input data.
+        context: Context
+            The context of the causal discovery problem.
+        """
+        X = data_df.to_numpy()
+        self.context = context
+
+        # Check acyclicity condition on included_edges
+        self._dag_check_included_edges()
+        self.order_constraints = self._included_edges_order_constraints()
+
+        # Inference of the causal order.
+        A_dense, order = self.top_order(X)
+        self.order_ = order
+
+        # Inference of the causal graph via pruning
+        A = self.prune(X, A_dense)
+        self.graph_ = nx.from_numpy_array(A, create_using=nx.DiGraph)
```

Comment:
See here, @francescomontagna you are converting stuff to a numpy array, but then throwing away the node labels. 

Good catch on the example. We should add a separate unit-test function to verify that the labels before/after fitting match that of what is fed in (this is prolly something basic that should be present for all algos).

Reply:
I added the summary paragraph and fixed the bullet point list.
I also fixed the behaviour causing the loss of information and wrote a unit test accordingly. See e.g. https://github.com/py-why/dodiscover/pull/129/files#r1211564581 . This can be easily re-used for any other method.